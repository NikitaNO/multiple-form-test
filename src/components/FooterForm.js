import React from 'react';

const FooterForm = props => {
    const { isBack, previousPage } = props;
    return (
      <div className="footer-form">
          <div className="next-wrapper"></div>
          {isBack &&
          <button type="button" className="previous" onClick={previousPage}>
              Previous
          </button>
          }
          <button type="submit" className="next">
              Next
              <img src="/src/img/right-arrow.svg" alt=""/>
          </button>
      </div>
    )
}

export default FooterForm;