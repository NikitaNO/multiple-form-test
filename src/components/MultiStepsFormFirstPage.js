import React from 'react';
import {Field, reduxForm} from 'redux-form';
import { renderField } from '../helpers/formComponets'
import FooterForm from './FooterForm';



@reduxForm({
  form: 'multi',
  destroyOnUnmount: false,
  forceUnregisterOnUnmount: true,
  // validate,
})
export default class MultiStepsFormFirstPage extends React.Component{
  render() {
    const { handleSubmit } = this.props;
    return (
      <form onSubmit={handleSubmit} className="first-form">
        <Field
          name="email"
          type="email"
          component={renderField}
          label="Email"
        />
        <Field
          name="password"
          type="password"
          component={renderField}
          label="Password"
        />
        <Field
          name="confirm"
          type="password"
          component={renderField}
          label="Confirm Password"
        />
        <FooterForm />
      </form>
    );
  }
};

function validate(values) {
  const errors = {};
  if (!values.email) {
    errors.email = 'Email Is Required';
  } else if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(values.email)) {
    errors.email = 'Invalid email address';
  }

  // Passwords
  if (!values.password) {
    errors.password = 'Password Is Required';
  } else if (values.password.length < 6) {
    errors.password = 'Password length must be more then 5 symbols';
  }
  if (values.confirm !== values.password) {
    errors.confirm = 'Passwords must be compared';
  }
  return errors;
}


