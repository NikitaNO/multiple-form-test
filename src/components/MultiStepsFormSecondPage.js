import React from 'react';
import {Field, reduxForm} from 'redux-form';
import { renderDate, renderError, renderSelector } from '../helpers/formComponets';
import FooterForm from './FooterForm';
import moment from 'moment';

@reduxForm({
  form: 'multi',
  destroyOnUnmount: false,
  forceUnregisterOnUnmount: true,
  validate,
})
export default class MultiStepsFormSecondPage extends React.Component {
  render() {
    const {handleSubmit, previousPage} = this.props;
    return (
      <form className="second-form" onSubmit={handleSubmit}>
        <label className="main-label">DATE OF BIRTH</label>
        <div className="date-wrapper">
          <Field
            name="birthdayDay"
            type="text"
            component={renderDate}
            placeholder="DD"
          />
          <Field
            name="birthdayMonth"
            type="number"
            component={renderDate}
            placeholder="MM"
          />
          <Field
            name="birthdayYear"
            type="number"
            component={renderDate}
            placeholder="YYYY"
          />
        </div>
        <div className="fixed-div">
          <label className="main-label">GENDER</label>
          <div className="field-wrapper-gender">
            <label className="label-gender" onClick={() => classToggle(0)}>
              <Field name="sex" component="input" type="radio" value="male"/>
              {' '}
              Male
            </label>
            <label className="label-gender label-gender-second" onClick={() => classToggle(1)}>
              <Field name="sex" component="input" type="radio" value="female"/>
              {' '}
              Female
            </label>
            <label className="label-gender label-gender-third" onClick={() => classToggle(2)}>
              <Field name="sex" component="input" type="radio" value="unspecified"/>
              {' '}
              Unspecified
            </label>
          </div>
          <Field name="sex" component={renderError}/>
        </div>
        <Field name="how_hear_about_us" component={renderSelector} options={["From media", "From friends"]}/>
        <FooterForm
          isBack={true}
          previousPage={previousPage}
        />
      </form>
    );
  }
};





const classToggle = (i) => {
  const menu = document.querySelectorAll('.label-gender');
  for (let cnt = 0; cnt < menu.length; cnt++) {
    menu[cnt].classList.remove('active');
    menu[cnt].classList.remove('active-first');
    menu[cnt].classList.remove('active-third');
  }
  switch (i) {
    case 0:
      menu[i].classList.add('active-first');
      break;
    case 1:
      menu[i].classList.add('active');
      break;
    case 2:
      menu[i].classList.add('active-third');
  }
}

function validate(values) {
  const errors = {};
  // Birthday day
  if (!values.birthdayDay) {
    errors.birthdayDay = 'Required';
  } else if (values.birthdayDay > 31 || values.birthdayDay < 1) {
    errors.birthdayDay = 'Invalid';
  }

  // Birthday month
  if (!values.birthdayMonth) {
    errors.birthdayMonth = 'Required';
  } else if (values.birthdayMonth > 12 || values.birthdayMonth < 1) {
    errors.birthdayMonth = 'Invalid';
  } else if (values.birthdayMonth.length < 2) {
    errors.birthdayMonth = 'Invalid';
  }

  // Birthday year
  if (!values.birthdayYear) {
    errors.birthdayYear = 'Required';
  } else if (values.birthdayYear.length != 4) {
    errors.birthdayYear = 'Invalid';
  } else {
    const userDate = new Date(`${values.birthdayYear}-${values.birthdayMonth}-${values.birthdayDay}`);
    const isAdult = moment(userDate, "YYYYMMDD").fromNow().split(' ')[0];
    if (parseInt(isAdult) < 18) {
      errors.birthdayYear = 'Not Adult';
    } else if (isNaN(parseInt(isAdult))) {
      errors.birthdayYear = 'You are not born';
    }
  }

  // Sex
  if (!values.sex) {
    errors.sex = 'Required';
  }
  return errors;
}

