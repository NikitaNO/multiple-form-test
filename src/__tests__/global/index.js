import React from 'react';
import renderer from 'react-test-renderer';
import FooterForm from '../../components/FooterForm';
import MultiStepsFormFirstPage from '../../components/MultiStepsFormFirstPage';
import MultiStepsFormSecondPage from '../../components/MultiStepsFormSecondPage';
import { shallow } from 'enzyme';

test('is Footer', () => {
    const footer = renderer.create(
      <FooterForm />
    );
    let footer_tree = footer.toJSON();
    expect(footer_tree).toMatchSnapshot();
});

it('submit first form', () => {

    const wrapper = shallow(
      <MultiStepsFormFirstPage />
    );

    const footer = shallow(
      <FooterForm />
    );

    expect(wrapper).toMatchSnapshot();
    wrapper.find('input[name="email"]').valueOf('test@mail.com');
    wrapper.find('input[name="password"]').valueOf('123456');
    wrapper.find('input[name="confirm"]').valueOf('123456');
    footer.find('.next').simulate('click');
    expect(wrapper).toMatchSnapshot();
});

it('submit second form', () => {

    const wrapper = shallow(
      <MultiStepsFormSecondPage />
    );

    const footer = shallow(
      <FooterForm />
    );

    expect(wrapper).toMatchSnapshot();
    wrapper.find('input[name="birthdayDay"]').valueOf('11');
    wrapper.find('input[name="birthdayMonth"]').valueOf('12');
    wrapper.find('input[name="birthdayYear"]').valueOf('1991');
    wrapper.find('input[name="sex"]').valueOf('Female');
    footer.find('.next').simulate('click');
    expect(wrapper).toMatchSnapshot();
});