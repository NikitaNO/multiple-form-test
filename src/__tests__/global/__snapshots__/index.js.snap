// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`is Footer 1`] = `
<div
  className="footer-form"
>
  <button
    className="next"
    type="submit"
  >
    Next
    <img
      alt=""
      src={Object {}}
    />
  </button>
</div>
`;

exports[`submit first form 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Connect(Form(MultiStepsFormFirstPage))
    destroyOnUnmount={false}
    enableReinitialize={false}
    forceUnregisterOnUnmount={true}
    form="multi"
    getFormState={[Function]}
    initialValues={undefined}
    keepDirtyOnReinitialize={false}
    persistentSubmitErrors={false}
    pure={true}
    shouldAsyncValidate={[Function]}
    shouldValidate={[Function]}
    touchOnBlur={true}
    touchOnChange={false}
    validate={[Function]}
/>,
  "nodes": Array [
    <Connect(Form(MultiStepsFormFirstPage))
      destroyOnUnmount={false}
      enableReinitialize={false}
      forceUnregisterOnUnmount={true}
      form="multi"
      getFormState={[Function]}
      initialValues={undefined}
      keepDirtyOnReinitialize={false}
      persistentSubmitErrors={false}
      pure={true}
      shouldAsyncValidate={[Function]}
      shouldValidate={[Function]}
      touchOnBlur={true}
      touchOnChange={false}
      validate={[Function]}
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ReduxForm />,
      "_debugID": 6,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ReduxForm {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {},
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Connect(Form(MultiStepsFormFirstPage))
          destroyOnUnmount={false}
          enableReinitialize={false}
          forceUnregisterOnUnmount={true}
          form="multi"
          getFormState={[Function]}
          initialValues={undefined}
          keepDirtyOnReinitialize={false}
          persistentSubmitErrors={false}
          pure={true}
          shouldAsyncValidate={[Function]}
          shouldValidate={[Function]}
          touchOnBlur={true}
          touchOnChange={false}
          validate={[Function]}
/>,
        "_debugID": 7,
        "_renderedOutput": <Connect(Form(MultiStepsFormFirstPage))
          destroyOnUnmount={false}
          enableReinitialize={false}
          forceUnregisterOnUnmount={true}
          form="multi"
          getFormState={[Function]}
          initialValues={undefined}
          keepDirtyOnReinitialize={false}
          persistentSubmitErrors={false}
          pure={true}
          shouldAsyncValidate={[Function]}
          shouldValidate={[Function]}
          touchOnBlur={true}
          touchOnChange={false}
          validate={[Function]}
/>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ReduxForm />,
}
`;

exports[`submit first form 2`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Connect(Form(MultiStepsFormFirstPage))
    destroyOnUnmount={false}
    enableReinitialize={false}
    forceUnregisterOnUnmount={true}
    form="multi"
    getFormState={[Function]}
    initialValues={undefined}
    keepDirtyOnReinitialize={false}
    persistentSubmitErrors={false}
    pure={true}
    shouldAsyncValidate={[Function]}
    shouldValidate={[Function]}
    touchOnBlur={true}
    touchOnChange={false}
    validate={[Function]}
/>,
  "nodes": Array [
    <Connect(Form(MultiStepsFormFirstPage))
      destroyOnUnmount={false}
      enableReinitialize={false}
      forceUnregisterOnUnmount={true}
      form="multi"
      getFormState={[Function]}
      initialValues={undefined}
      keepDirtyOnReinitialize={false}
      persistentSubmitErrors={false}
      pure={true}
      shouldAsyncValidate={[Function]}
      shouldValidate={[Function]}
      touchOnBlur={true}
      touchOnChange={false}
      validate={[Function]}
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ReduxForm />,
      "_debugID": 6,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ReduxForm {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {},
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Connect(Form(MultiStepsFormFirstPage))
          destroyOnUnmount={false}
          enableReinitialize={false}
          forceUnregisterOnUnmount={true}
          form="multi"
          getFormState={[Function]}
          initialValues={undefined}
          keepDirtyOnReinitialize={false}
          persistentSubmitErrors={false}
          pure={true}
          shouldAsyncValidate={[Function]}
          shouldValidate={[Function]}
          touchOnBlur={true}
          touchOnChange={false}
          validate={[Function]}
/>,
        "_debugID": 7,
        "_renderedOutput": <Connect(Form(MultiStepsFormFirstPage))
          destroyOnUnmount={false}
          enableReinitialize={false}
          forceUnregisterOnUnmount={true}
          form="multi"
          getFormState={[Function]}
          initialValues={undefined}
          keepDirtyOnReinitialize={false}
          persistentSubmitErrors={false}
          pure={true}
          shouldAsyncValidate={[Function]}
          shouldValidate={[Function]}
          touchOnBlur={true}
          touchOnChange={false}
          validate={[Function]}
/>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ReduxForm />,
}
`;

exports[`submit second form 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Connect(Form(Connect(MultiStepsFormSecondPage)))
    destroyOnUnmount={false}
    enableReinitialize={false}
    forceUnregisterOnUnmount={true}
    form="multi"
    getFormState={[Function]}
    initialValues={undefined}
    keepDirtyOnReinitialize={false}
    persistentSubmitErrors={false}
    pure={true}
    shouldAsyncValidate={[Function]}
    shouldValidate={[Function]}
    touchOnBlur={true}
    touchOnChange={false}
    validate={[Function]}
/>,
  "nodes": Array [
    <Connect(Form(Connect(MultiStepsFormSecondPage)))
      destroyOnUnmount={false}
      enableReinitialize={false}
      forceUnregisterOnUnmount={true}
      form="multi"
      getFormState={[Function]}
      initialValues={undefined}
      keepDirtyOnReinitialize={false}
      persistentSubmitErrors={false}
      pure={true}
      shouldAsyncValidate={[Function]}
      shouldValidate={[Function]}
      touchOnBlur={true}
      touchOnChange={false}
      validate={[Function]}
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ReduxForm />,
      "_debugID": 10,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ReduxForm {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {},
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 3,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Connect(Form(Connect(MultiStepsFormSecondPage)))
          destroyOnUnmount={false}
          enableReinitialize={false}
          forceUnregisterOnUnmount={true}
          form="multi"
          getFormState={[Function]}
          initialValues={undefined}
          keepDirtyOnReinitialize={false}
          persistentSubmitErrors={false}
          pure={true}
          shouldAsyncValidate={[Function]}
          shouldValidate={[Function]}
          touchOnBlur={true}
          touchOnChange={false}
          validate={[Function]}
/>,
        "_debugID": 11,
        "_renderedOutput": <Connect(Form(Connect(MultiStepsFormSecondPage)))
          destroyOnUnmount={false}
          enableReinitialize={false}
          forceUnregisterOnUnmount={true}
          form="multi"
          getFormState={[Function]}
          initialValues={undefined}
          keepDirtyOnReinitialize={false}
          persistentSubmitErrors={false}
          pure={true}
          shouldAsyncValidate={[Function]}
          shouldValidate={[Function]}
          touchOnBlur={true}
          touchOnChange={false}
          validate={[Function]}
/>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ReduxForm />,
}
`;

exports[`submit second form 2`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Connect(Form(Connect(MultiStepsFormSecondPage)))
    destroyOnUnmount={false}
    enableReinitialize={false}
    forceUnregisterOnUnmount={true}
    form="multi"
    getFormState={[Function]}
    initialValues={undefined}
    keepDirtyOnReinitialize={false}
    persistentSubmitErrors={false}
    pure={true}
    shouldAsyncValidate={[Function]}
    shouldValidate={[Function]}
    touchOnBlur={true}
    touchOnChange={false}
    validate={[Function]}
/>,
  "nodes": Array [
    <Connect(Form(Connect(MultiStepsFormSecondPage)))
      destroyOnUnmount={false}
      enableReinitialize={false}
      forceUnregisterOnUnmount={true}
      form="multi"
      getFormState={[Function]}
      initialValues={undefined}
      keepDirtyOnReinitialize={false}
      persistentSubmitErrors={false}
      pure={true}
      shouldAsyncValidate={[Function]}
      shouldValidate={[Function]}
      touchOnBlur={true}
      touchOnChange={false}
      validate={[Function]}
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ReduxForm />,
      "_debugID": 10,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ReduxForm {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {},
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 3,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Connect(Form(Connect(MultiStepsFormSecondPage)))
          destroyOnUnmount={false}
          enableReinitialize={false}
          forceUnregisterOnUnmount={true}
          form="multi"
          getFormState={[Function]}
          initialValues={undefined}
          keepDirtyOnReinitialize={false}
          persistentSubmitErrors={false}
          pure={true}
          shouldAsyncValidate={[Function]}
          shouldValidate={[Function]}
          touchOnBlur={true}
          touchOnChange={false}
          validate={[Function]}
/>,
        "_debugID": 11,
        "_renderedOutput": <Connect(Form(Connect(MultiStepsFormSecondPage)))
          destroyOnUnmount={false}
          enableReinitialize={false}
          forceUnregisterOnUnmount={true}
          form="multi"
          getFormState={[Function]}
          initialValues={undefined}
          keepDirtyOnReinitialize={false}
          persistentSubmitErrors={false}
          pure={true}
          shouldAsyncValidate={[Function]}
          shouldValidate={[Function]}
          touchOnBlur={true}
          touchOnChange={false}
          validate={[Function]}
/>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ReduxForm />,
}
`;
